Bluetooth: Audio: Add TMAS and two TMAP samples

sp3030 opened the pull request:
Add implementation of the Telephony and Media Audio Service, as well as two sample applications.

tmap_central reflects a smartphone implementing the Unicast Media Sender and Call Gateway TMAP roles.

tmap_peripheral reflects an earbud implementing the Unicast Media Receiver and Call Terminal TMAP roles.

Upon connection, tmap_central starts an audio stream using CAP Initiator APIs.

CCP, MCP and VCP are discovered and used to send example commands.

Future improvements: 2-earbud support, add TMAP Broadcast roles, update with new CAP Acceptor/Commander APIs as they become available

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/49532
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/49533
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/49534
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/49535
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/49539

Thalley commented on tests/bsim/bluetooth/audio/src/tmas_test.c:
```diff
@@ -0,0 +1,120 @@
+
+static uint8_t tmap_addata[] = {
+	BT_UUID_16_ENCODE(BT_UUID_TMAS_VAL), /* TMAS UUID */
+	(BT_TMAP_ROLE_UNICAST_MEDIA_RECEIVER) | (BT_TMAP_ROLE_CALL_TERMINAL), 0x00, /* TMAP Role */
```
```suggestion
	BT_TMAP_ROLE_UNICAST_MEDIA_RECEIVER | BT_TMAP_ROLE_CALL_TERMINAL, 0x00, /* TMAP Role */
```

Thalley commented on tests/bsim/bluetooth/audio/src/tmas_test.c:
```diff
@@ -0,0 +1,120 @@
+	err = bt_tmas_init();
+	if (err != 0) {
+		return;
+	}
+	bt_tmas_register_role((BT_TMAP_ROLE_CALL_TERMINAL) | (BT_TMAP_ROLE_UNICAST_MEDIA_RECEIVER));
```
```suggestion
	}
	
	bt_tmas_register_role(BT_TMAP_ROLE_CALL_TERMINAL | BT_TMAP_ROLE_UNICAST_MEDIA_RECEIVER);
```

Thalley commented on tests/bsim/bluetooth/audio/test_scripts/tmas.sh:
The BSIM shell scripts were recently refactored to not define the `Execute` function etc. but rather use a common shell file. Please have a look at the other test scripts and modify this to be similar to them

Thalley reviewed (changes requested):
I think this is a quite good start for the implementation. I've added some comments that should be addressed

sp3030 commented:
> I think this is a quite good start for the implementation. I've added some comments that should be addressed

Thank you very much for the feedback, will get right on it.

Thalley commented on include/zephyr/bluetooth/audio/tmap.h:
```diff
@@ -0,0 +1,65 @@
+ */
+int bt_tmap_discover(struct bt_conn *conn, struct bt_tmap_cb *tmap_cb);
+
+/**
+ * @brief Register TMAP role(s)
+ *
+ * @param role     TMAP role(s).
+ *
+ */
+void bt_tmap_register_role(enum bt_tmap_role role);
```
Duplicate of `bt_tmap_register`?

Thalley commented on include/zephyr/bluetooth/audio/tmap.h:
```diff
@@ -0,0 +1,65 @@
+/**
+ * @brief Adds TMAS instance to database and sets the received TMAP role(s).
+ *
+ * @param role TMAP role of the device
```
Should better indicate that `role` can contain multiple `bt_tmap_role` values (i.e. that it is a bitfield)

Thalley commented on include/zephyr/bluetooth/audio/tmap.h:
```diff
@@ -0,0 +1,65 @@
+	 *  This callback notifies the application about the value of the
+	 *  TMAP Role characteristic on the peer.
+	 *
+	 *  @param role		Peer TMAP role.
+	 *  @param conn     Pointer to the connection
+	 *  @param err      0 if success, -EINVAL if TMAS was not discovered.
```
```suggestion
	 *  @param role     Peer TMAP role(s).
	 *  @param conn     Pointer to the connection
	 *  @param err      0 if success, -EINVAL if TMAS was not discovered.
```

Thalley replied:
`EOPNOTSUPP` would also be better than `EINVAL` if we really want to return a errno value here. Alternatively we can just return the ATT error from the server

Thalley commented on include/zephyr/bluetooth/audio/tmap.h:
```diff
@@ -0,0 +1,65 @@
+	 *  @param conn     Pointer to the connection
+	 *  @param err      0 if success, -EINVAL if TMAS was not discovered.
+	 */
+	void (*tmap_discovery_complete)(enum bt_tmap_role role, struct bt_conn *conn, int err);
```
```suggestion
	void (*discovery_complete)(enum bt_tmap_role role, struct bt_conn *conn, int err);
```

Thalley commented on samples/bluetooth/tmap_central/src/cap_initiator.c:
```diff
@@ -0,0 +1,513 @@
+
+static struct k_work_delayable audio_send_work;
+
+static struct bt_cap_stream unicast_streams[CONFIG_BT_BAP_UNICAST_CLIENT_ASE_SNK_COUNT +
+										CONFIG_BT_BAP_UNICAST_CLIENT_ASE_SRC_COUNT];
```
Some indentation went wrong here. Please note that tabs in Zephyr are 8 spaces :)

Thalley replied:
This seems to be a general issue in this commit :)

Thalley commented on subsys/bluetooth/audio/Kconfig.tmap:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config BT_TMAP
+	bool "Telephone and Media Audio Profile [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  Enabling this will enable TMAP.
```
Should we add dependencies for the other profiles here? 
A TMAP device needs to be either a CAP Initiator or a CAP Acceptor, right?

Thalley commented on subsys/bluetooth/audio/tmap.c:
```diff
@@ -0,0 +1,154 @@
+
+	if (!attr) {
+		(void)memset(params, 0, sizeof(*params));
+		if (tmas_found == false) {
```
```suggestion
		if (!tmas_found) {
```

Thalley commented on subsys/bluetooth/audio/tmap.c:
```diff
@@ -0,0 +1,154 @@
+		if (tmas_found == false) {
+			/* TMAS not found on peer */
+			if (cb->tmap_discovery_complete) {
+				cb->tmap_discovery_complete(0, conn, -EINVAL);
```
Suggest to either use `BT_ATT_ERR_ATTRIBUTE_NOT_FOUND` or `EOPNOTSUPP` instead of `EINVAL`

Thalley commented on subsys/bluetooth/audio/tmap.c:
```diff
@@ -0,0 +1,154 @@
+			if (cb->tmap_discovery_complete) {
+				cb->tmap_discovery_complete(0, conn, -EINVAL);
+			}
+		}
+		tmas_found = false;
```
```suggestion
		}
		
		tmas_found = false;
```

Thalley commented on subsys/bluetooth/audio/tmap.c:
```diff
@@ -0,0 +1,154 @@
+
+LOG_MODULE_REGISTER(bt_tmap, CONFIG_BT_TMAP_LOG_LEVEL);
+
+static uint16_t tmap_role;
+static struct bt_tmap_cb *cb;
+static bool tmas_found;
+
+static struct bt_uuid_16 uuid = BT_UUID_INIT_16(0);
+static struct bt_gatt_discover_params discover_params;
+static struct bt_gatt_read_params read_params;
```
If we want to support multiple discovery procedures in parallel (i.e. for multiple connections), then these values need to be in an array of some sort.

Thalley commented on subsys/bluetooth/audio/tmap.c:
```diff
@@ -0,0 +1,154 @@
+
+	if (cb->tmap_discovery_complete) {
+		cb->tmap_discovery_complete(peer_role, conn, err);
+	}
+	return BT_GATT_ITER_STOP;
```
```suggestion
	}
	
	return BT_GATT_ITER_STOP;
```

Thalley commented on subsys/bluetooth/audio/tmap.c:
```diff
@@ -0,0 +1,154 @@
+					struct bt_gatt_read_params *params,
+					const void *data, uint16_t length)
+{
+	enum bt_tmap_role peer_role = sys_get_le16(data);
```
We should not assume that the size of an `enum` can always hold a 16-bit value. 

Ideally we do something like 
```
uint16_t peer_role = sys_get_le16(data);

if (!peer_role_is_valid) {
	if (cb->tmap_discovery_complete) {
		cb->tmap_discovery_complete(0, conn, BT_ATT_ERR_VALUE_NOT_ALLOWED);
	}
} else {
	if (cb->tmap_discovery_complete) {
		cb->tmap_discovery_complete((enum bt_tmap_role)peer_role, conn, 0);
	}
}
```

We also should verify the length before we call `sys_get_le16` and should also check that `err == BT_ATT_ERR_SUCCESS` before attempting to extract any data

Thalley commented on tests/bsim/bluetooth/audio/src/main.c:
```diff
@@ -56,6 +58,8 @@ bst_test_install_t test_installers[] = {
 	test_has_client_install,
 	test_ias_install,
 	test_ias_client_install,
+	test_tmap_install,
+	test_tmap_client_install,
```
Suggest to rename these to `tmap_central` and `tmap_peripheral` or `tmap_client` and `tmap_server`

Thalley commented on tests/bsim/bluetooth/audio/src/tmap_client_test.c:
```diff
@@ -0,0 +1,194 @@
+	}
+
+	peer_tmap_role = net_buf_simple_pull_le16(&tmas_svc_data);
+	if (!(sys_le16_to_cpu(peer_tmap_role) & BT_TMAP_ROLE_UMR)) {
```
```suggestion
	if (!(peer_tmap_role & BT_TMAP_ROLE_UMR)) {
```

Thalley commented on tests/bsim/bluetooth/audio/src/tmap_client_test.c:
```diff
@@ -0,0 +1,194 @@
+	.recv = scan_recv,
+};
+
+static void connected_cb(struct bt_conn *conn, uint8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+	if (err > 0) {
+		FAIL("Failed to connect to %s (err %u)\n", addr, err);
+		return;
+	}
+	printk("TMAP Central Connected to %s\n", addr);
+	SET_FLAG(flag_tmap_connected);
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected_cb
+};
```
Can use the commonly defined `connected` callback and flag

Thalley commented on tests/bsim/bluetooth/audio/src/tmap_test.c:
```diff
@@ -0,0 +1,119 @@
+	BT_DATA(BT_DATA_SVC_DATA16, tmap_addata, ARRAY_SIZE(tmap_addata)),
+};
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+	if (err > 0) {
+		FAIL("Failed to connect to %s (err %u)\n", addr, err);
+		return;
+	}
+	printk("TMAP Peripheral Connected to %s\n", addr);
+	SET_FLAG(flag_tmap_connected);
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected
+};
```
Can use the common callback and flag (see the other tests)

Thalley commented on include/zephyr/bluetooth/audio/tmap.h:
```diff
@@ -55,11 +55,12 @@ int bt_tmap_register(enum bt_tmap_role role);
 
 /**
- * @brief Register TMAP role(s)
+ * @brief Register one or multiple TMAP roles dynamically.
```
```suggestion
 * @brief Set one or multiple TMAP roles dynamically.
```

Thalley commented on samples/bluetooth/tmap_central/src/cap_initiator.c:
```diff
@@ -21,20 +21,20 @@
 
 NET_BUF_POOL_FIXED_DEFINE(tx_pool, CONFIG_BT_BAP_UNICAST_CLIENT_ASE_SNK_COUNT,
-						CONFIG_BT_ISO_TX_MTU + BT_ISO_CHAN_SEND_RESERVE, 8, NULL);
+			CONFIG_BT_ISO_TX_MTU + BT_ISO_CHAN_SEND_RESERVE, 8, NULL);
```
There still seems to be some general issues with the indentation :) Tabs are 8 spaces, and code should be indented to match previous lines (e.g. by `(`)

Thalley commented on subsys/bluetooth/audio/tmap.c:
```diff
@@ -22,63 +22,97 @@
+	/* Extract the TMAP role of the peer and inform application fo the value found */
+	uint16_t peer_role = sys_get_le16(data);
+
+	if (peer_role & TMAP_ALL_ROLES) {
+		if (cb->discovery_complete) {
+			cb->discovery_complete((enum bt_tmap_role)peer_role, conn, 0);
+		}
```
If `peer_role` is `0xFFFF` then it's an invalid role, but isn't caught by this check :)

Thalley commented on subsys/bluetooth/audio/tmap.c:
```diff
@@ -22,63 +22,97 @@
 	}
+
+	/* Check data length */
+	if (length != sizeof(uint16_t)) {
+		if (cb->discovery_complete) {
+			cb->discovery_complete(0, conn, BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+		}
+
+		return BT_GATT_ITER_STOP;
+	}
+
+	/* Extract the TMAP role of the peer and inform application fo the value found */
+	uint16_t peer_role = sys_get_le16(data);
```
Declare `peer_role` at top of scope and do

```suggestion
	if (length != sizeof(peer_role)) {
		if (cb->discovery_complete) {
			cb->discovery_complete(0, conn, BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
		}

		return BT_GATT_ITER_STOP;
	}

	/* Extract the TMAP role of the peer and inform application fo the value found */
	peer_role = sys_get_le16(data);
```

Thalley reviewed (changes requested):
A few comments. 

Please also follow the Contribution Workflow as defined by https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow and squash the commits to a single commit (instead of adding "fixing" commits). 

Could also consider splitting this PR in the following commits:
1) Add TMAP implementation
2) Add TMAP tests
3) Add TMAP samples

sp3030 committed:
Bluetooth: Audio: Add TMAS and two TMAP samples

Add implementation of the Telephony and Media Audio Service, as well as
two sample applications.

tmap_central reflects a smartphone implementing the Unicast Media Sender
and Call Gateway TMAP roles.

tmap_peripheral reflects an earbud implementing the Unicast Media
Receiver and Call Terminal TMAP roles.

Upon connection, tmap_central starts an audio stream using CAP Initiator
APIs.

CCP, MCP and VCP are discovered and used to send example commands.

Future improvements: 2-earbud support, add TMAP Broadcast roles,
update with new CAP Acceptor/Commander APIs as they become available

Signed-off-by: Silviu Petria <silviu.petria@nxp.com>

Thalley reviewed (approved):
I don't have any other blocking comments - Nice work!

MariuszSkamra reviewed (approved):
LGTM
